### ISSUE TITLE
Setting and retrieving user in same request

### ISSUE THREAD

**Issue by mmzoo** [hassox/warden#103]
### TL;DR

```ruby
# config/initializers/warden.rb
class Warden::SessionSerializer
  def deserialize(attributes)
    User.new attributes
  end
end

# some_controller.rb
warden.set_user id: 42, name: 'Bob'
warden.user  # => A Hash, not a User.
```

### More elaborate:

When I set the user in my SessionsController, I pass in a `Hash` with user attributes. A simple `Warden::SessionSerializer`  runs `User.new(attributes)` to turn that into an instance each time I access `warden.user`. After login I redirect the user and have the user available as an `User` instance. So far so good.

Now I wish to mutate the user but don't want to have to wait one request for the changes to take effect. I would like to set the user and be able to retrieve it immediately after setting it. How would I go about that?

Because if I set the user as a `Hash`, a Hash is what I get out (according to the code [to set](https://github.com/hassox/warden/blob/e43d3ed7a1d8a8915b39e0ebbd21c733beca00d5/lib/warden/proxy.rb#L170) and [retrieve](https://github.com/hassox/warden/blob/e43d3ed7a1d8a8915b39e0ebbd21c733beca00d5/lib/warden/proxy.rb#L209-L210) a user, no immediate serialization takes place). If I set it as a `User` instance, it works for this request but the deserializer is going to be confused when not confronted with a `Hash` at the next request.

On logout, [the code](https://github.com/hassox/warden/blob/e43d3ed7a1d8a8915b39e0ebbd21c733beca00d5/lib/warden/proxy.rb#L261-L265) takes effect immediately by (understandably) cleaning out the `@users` variable **and** the `session_serializer`. I'm wondering why setting and retrieving the user doesn't behave similarly. Is that intentional?

Now, it does turn out that I can do this in my middleware to achieve the desired effect, but it feels really awkward:

```ruby
warden.user.attributes = something: :new         # <- Affects current request
warden.session_serializer.store user.attributes  # <- Affects subsequent requests
```

Is it okay to do that or did I miss something? I'd appreciate any pointers.

(Maybe just to clarify my use case: I have a middleware that checks the user's single-sign-on session with a remote server and is supposed to mutate the user depending on what that SSO server says about the user.)

Thank you for your time! Warden is wonderful otherwise :)

**Comment by mmzoo** [comment]
PS: I assume I'm supposed to do this, right?

```ruby
env['warden'].set_user(user, :store => false)   # <- Affects current request
env['warden'].set_user(user.attributes)         # <- Affects subsequent requests
```

**Comment by mmzoo** [comment]
It's not ideal, but it works.

**Comment by mmzoo** [comment]
Ok, but [this](https://github.com/hassox/warden/blob/e43d3ed7a1d8a8915b39e0ebbd21c733beca00d5/lib/warden/strategies/base.rb#L116) doesn't work because at login I don't have a user at hand yet. 

Say I have this setup:

```ruby
Warden::Manager.serialize_into_session do |user_id|
  user = ::Customers.find_by_id(user_id)
  user.attributes
end

Warden::Manager.serialize_from_session do |serialized|
  ::Customer.new serialized
end
```
So I would normally run `success! 42` and that's enough to have a `Customer` instance available at `warden.user` at the next request. But I need the instance in the same request now.

I really feel bad duplicating the serializing logic. As a workaround, is there a way to call the `Warden::Manager.serialize_into_session` and then the `Warden::Manager.serialize_from_session` logic manually so I can pass the result to `success!`?

**Comment by mmzoo** [comment]
OK, I found a workaround again:

```ruby
user_id = warden.authenticate! :password
serialized = warden.session_serializer.serialize(user_id)
deserialized = warden.session_serializer.deserialize(serialized)
```
